File: src/lib/email.ts
import type { EnvLike } from "./env";

/**
 * Transactional email, behind one function.
 *
 * Resend's REST API over `fetch` rather than its SDK: M0 is barred from
 * dependency changes and this is a single POST. Everything provider-specific
 * lives in this file, so swapping to Postmark or SES later is one file and no
 * caller changes. This is the same seam the M2 verification and password-reset
 * mails will use.
 */

const RESEND_ENDPOINT = "https://api.resend.com/emails";

/** Give up rather than hold the visitor's browser open indefinitely. */
const SEND_TIMEOUT_MS = 8_000;

export type EmailMessage = {
  to: string;
  from: string;
  subject: string;
  text: string;
  /** Where a reply from the monitored inbox should go — the lead, not us. */
  replyTo?: string;
};

export class EmailNotConfiguredError extends Error {
  constructor(missing: string[]) {
    super(`Email is not configured: missing ${missing.join(", ")}`);
    this.name = "EmailNotConfiguredError";
  }
}

export class EmailSendError extends Error {
  constructor(
    message: string,
    readonly status?: number,
  ) {
    super(message);
    this.name = "EmailSendError";
  }
}

export type EmailConfig = {
  apiKey: string;
  from: string;
  to: string;
};

/**
 * Reads and validates configuration at call time rather than at module load,
 * so a missing variable surfaces as a handled 502 on one route instead of a
 * boot failure that takes the whole marketing site down.
 */
export function readEmailConfig(env: EnvLike = process.env): EmailConfig {
  const apiKey = env.RESEND_API_KEY?.trim();
  const from = env.CONTACT_FROM_EMAIL?.trim();
  const to = env.CONTACT_TO_EMAIL?.trim();

  const missing = [
    !apiKey && "RESEND_API_KEY",
    !from && "CONTACT_FROM_EMAIL",
    !to && "CONTACT_TO_EMAIL",
  ].filter((name): name is string => Boolean(name));

  if (missing.length > 0) throw new EmailNotConfiguredError(missing);

  return { apiKey: apiKey!, from: from!, to: to! };
}

export async function sendEmail(message: EmailMessage, config: EmailConfig): Promise<string> {
  let response: Response;

  try {
    response = await fetch(RESEND_ENDPOINT, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${config.apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        from: message.from,
        to: [message.to],
        subject: message.subject,
        text: message.text,
        ...(message.replyTo ? { reply_to: [message.replyTo] } : {}),
      }),
      signal: AbortSignal.timeout(SEND_TIMEOUT_MS),
    });
  } catch (cause) {
    throw new EmailSendError(
      cause instanceof Error && cause.name === "TimeoutError"
        ? `Email provider did not respond within ${SEND_TIMEOUT_MS}ms`
        : `Email provider unreachable: ${cause instanceof Error ? cause.message : String(cause)}`,
    );
  }

  if (!response.ok) {
    // Body is provider diagnostics, never the visitor's message.
    const detail = await response.text().catch(() => "");
    throw new EmailSendError(
      `Email provider returned ${response.status}: ${detail.slice(0, 500)}`,
      response.status,
    );
  }

  const body = (await response.json().catch(() => ({}))) as { id?: string };
  return body.id ?? "";
}

File: services/api/soyl/infrastructure/email.py
"""Transactional email, behind one function.

The Python counterpart of `src/lib/email.ts`, deliberately the same shape: one
POST to Resend's REST API, no SDK, everything provider-specific in this file.
Swapping to Postmark or SES stays one file on each side.

Two senders exist because two runtimes send: the web app sends the contact-form
notification, the API sends verification and reset. They should hold *separate*
API keys so revoking one does not take out the other.
"""

from __future__ import annotations

import httpx

RESEND_ENDPOINT = "https://api.resend.com/emails"
SEND_TIMEOUT_SECONDS = 8.0


class EmailNotConfigured(Exception):
    """No API key. In local development this is normal, not a failure."""


class EmailSendFailed(Exception):
    def __init__(self, message: str, status: int | None = None) -> None:
        self.status = status
        super().__init__(message)


class EmailSender:
    """Sends, or deliberately does not.

    When unconfigured it raises `EmailNotConfigured` rather than silently
    succeeding — the caller decides whether that is fatal. For signup it is
    not: an account is still created and the verification link can be resent.
    """

    def __init__(self, *, api_key: str | None, from_address: str | None) -> None:
        self._api_key = api_key
        self._from = from_address

    @property
    def configured(self) -> bool:
        return bool(self._api_key and self._from)

    async def send(self, *, to: str, subject: str, text: str) -> str:
        if not self.configured:
            raise EmailNotConfigured

        try:
            async with httpx.AsyncClient(timeout=SEND_TIMEOUT_SECONDS) as client:
                response = await client.post(
                    RESEND_ENDPOINT,
                    headers={"Authorization": f"Bearer {self._api_key}"},
                    json={"from": self._from, "to": [to], "subject": subject, "text": text},
                )
        except (TimeoutError, httpx.HTTPError) as exc:
            raise EmailSendFailed(f"Email provider unreachable: {type(exc).__name__}") from exc

        if response.status_code >= 400:
            # Provider diagnostics only. The body of the message is never
            # logged and never included here.
            raise EmailSendFailed(
                f"Email provider returned {response.status_code}: {response.text[:500]}",
                response.status_code,
            )

        body: dict[str, str] = response.json()
        return body.get("id", "")


def verification_email(*, display_name: str | None, link: str) -> tuple[str, str]:
    greeting = f"Hi {display_name}," if display_name else "Hi,"
    return (
        "Confirm your email address",
        f"""{greeting}

Confirm your email address to finish setting up your SOYL account:

{link}

This link works once and expires in 24 hours. If you did not create an
account, you can ignore this email — nothing will happen.

— SOYL
""",
    )


def password_reset_email(*, display_name: str | None, link: str) -> tuple[str, str]:
    greeting = f"Hi {display_name}," if display_name else "Hi,"
    return (
        "Reset your password",
        f"""{greeting}

Someone asked to reset the password for this SOYL account. If it was you:

{link}

This link works once and expires in one hour. Using it will sign you out
everywhere else.

If it was not you, you can ignore this email. Your password has not changed
and nobody has been told whether this address has an account.

— SOYL
""",
    )
